torch/_library/infer_schema.py

```diff
@@ -318,3 +319,149 @@ def tuple_to_list(tuple_type: type[tuple]) -> type[list]:
         return list[type_args[0]]  # type: ignore[valid-type]
     else:
         return list[typing.Union[tuple(type_args)]]  # type: ignore[misc, return-value]
+
+
+# Below is an implementation of generating FunctionSchema from example values.
+# This is helpful for generating FunctionSchema for HigherOrderOperator, where
+# we don't have a function to inspect and each call of the higher order operator
+# would have different schema.
+@dataclass(frozen=True)
+class HopArgumentInfo:
+    # Could give a name to the operand by default it's empty string.
+    name: str
+    example_value: Any
+    # Provide an default_value
+    default_value: Any
+    # Whether this arugment gets mutated in the hop subgraph.
+    # For output, this should always be False
+    is_mutated: bool
+
+
+class HopArgumentInfoGen:
+    @staticmethod
+    def from_example(
+        example_value: Any,
+        *,
+        name: str = "",
+        default_value: Optional[Any],
+        is_mutated: bool = False,
+    ) -> HopArgumentInfo:
+        if default_value is not None:
+            assert type(example_value) == type(default_value)
+        return HopArgumentInfo(
+            name=name,
+            example_value=example_value,
+            default_value=default_value,
+            is_mutated=is_mutated,
+        )
+
+
+class CTypeGen:
+    convert_to_base_ty = {
+        int: torch._C.IntType.get(),
+        float: torch._C.FloatType.get(),
+        str: torch._C.StringType.get(),
+        bool: torch._C.BoolType.get(),
+    }
+
+    # should return torch._C.JitType but that annotation is busted
+    @staticmethod
+    def from_example(obj: Any) -> Any:
+        import torch
+
+        if isinstance(obj, torch.Tensor):
+            return torch._C.TensorType.get()
+        elif isinstance(obj, torch.fx.GraphModule):
+            return torch._C.AnyType.get()
+        elif isinstance(obj, torch.SymInt):
+            return (torch._C.SymIntType.get(),)
+        elif isinstance(obj, torch.SymBool):
+            return (torch._C.SymBoolType.get(),)
+        elif isinstance(obj, tuple):
+            return torch._C._create_tuple_type(
+                tuple(CTypeGen.from_example(arg) for arg in obj)
+            )
+        elif obj is None:
+            return torch._C.NoneType.get()
+        tp = type(obj)
+        if tp not in CTypeGen.convert_to_base_ty:
+            raise RuntimeError(f"unsupported type {tp}")
+        return CTypeGen.convert_to_base_ty[tp]
```

Comment:
I feel like there has to be a C++ function somewhere for this?
Can we try to parse an IValue out of the obj and then use the IValue's type?

And then if it fails, then we can do special cases?

Reply:
yeah, let me try to find it.